CMS-Enterprise/sbom-harbor
# feat(cpe-enrichment): add a purl to cpe enrichment source

## Summary

This PR will add a purl to cpe enrichment source and a new data set `construction` Provider to create the perl2cpe data set. 

### Added

Adds new Tasks and service that will create a store from data in the https://github.com/scanoss/purl2cpe GitHub repository, then use that data to enrich dependency type Packages with cpe(s) from their purl(s)  

## How to test

1. Set up test environment, specifically, MongoDB must be running
2. Run Snyk ingest provider: `./target/debug/harbor ingest -p snyk --debug`
3. Open Compass, select the `Package` collection and use this filter: `{ kind: "dependency", cpe: null }`. That will show all of the dependencies in MongoDB that have no CPE.
4. Run the purl2cpe dataset building Provider: `./target/debug/harbor construct purl2cpe`.  This builds the purl2cpe data set so we can use it to get CPEs for the purls of the dependencies we just ingested.
5. Using Compass, verify that a `Purl2Cpes` Collection exists in MongoDB and is populated.
6. Lastly, run the purl2cpe _enrichment_ provider: `./target/debug/harbor enrich -p purl2cpe`.  This process will go through the dependencies missing a cpe and add them based on the purl.
7. Once more, open Compass, select the `Package` collection and use this filter: `{ kind: "dependency", cpe: null }`. Now the filter should produce no results
8. Change the filter to: `{ kind: "dependency", cpe: { $ne: "unknown" } }` and verify that results exist.  Look for examples of the "all versions" cpe for a given Package.  I found one by filtering on: `{ purl: "pkg:maven/com.fasterxml.jackson.core/jackson-databind@2.10.3" }` for example.  Also, look to make sure the cpe's that have versions line up with the version of the Package.
